Connect Angular frontend to Express/Lambda backend

beechnut opened the pull request:
- Adds a service to the Angular app that connects to the Express API
- Adds CORS headers to Express

Closes #352

neilmb commented on api/src/index.ts:
```diff
@@ -3,6 +3,13 @@ import express, { Request, Response, NextFunction } from 'express';
 const port = 3000;
 
+app.use(function(req, res, next) {
+  res.setHeader('Access-Control-Allow-Origin', 'http://localhost:4200')
```
**question:** If you are adding per-environment controls for Angular in this PR, do we need to do something similar for the API to listen on a safe URL? Or is this part of the Express infrastructure that isn't run on Lambda? If so, maybe big, loud comments in this file about what is only part of the application for local development?

beechnut replied:
Trying to reply to this specific question had me realize that while the frontend and backend are connected in that they know how to locate each other locally, this PR does not yet cover deploying Angular at a path through the Express app. This seems like an essential part of this story, so I'll continue working on that.

beechnut replied:
I ended up having to set CORS headers to `'*'` to get the sandbox deployment to work, and I'm not sure why. Setting it to the frontend host (the S3 URL) didn't work. I wonder if this has something to do with the API Gateway, but the resources we're using via Claudia aren't visible to me in the sandbox environment.

neilmb commented on frontend/src/app/app.module.ts:
```diff
@@ -2,13 +2,18 @@ import { NgModule } from '@angular/core';
 import { AppComponent } from './app.component';
+import { TimezonesComponent } from './timezones/timezones.component';
+
+import { HttpClientModule } from '@angular/common/http';
```
**nitpick:** Is there a standard for how these imports are ordered in TS? I would expect to see this one with the other `@angular/...` import above the local imports.

beechnut replied:
I just added Prettier with a plugin that organizes TypeScript imports. Before I commit and push it, do you have an opinion on how you'd want Prettier to run? I'm thinking pre-commit hook, but I'm open to other ideas.

neilmb commented on frontend/src/app/timezone.service.spec.ts:
```diff
@@ -0,0 +1,16 @@
+    service = TestBed.inject(TimezoneService);
+  });
+
+  it('should be created', () => {
```
**question (if-minor):** Do we have documentation in our README on how to run these tests during local development? And are they running now in CI, or is that part of the planning work that we are still doing? If not, would adding the running of these tests to the existing setup be worth doing in this PR?

beechnut replied:
@neilmb Glad you're paying attention to this. I do think this is best done as part of another story. This one is just to connect the apps, but perhaps #289 would make a good home for this (even though they're unit tests).

neilmb commented on frontend/src/app/timezone.ts:
```diff
@@ -0,0 +1,9 @@
+export interface Timezone {
```
**question:** There is a very similar typedef in https://github.com/USDAForestService/NRM-Grants-Agreements/blob/development/api/src/index.ts#L10. For this, I'm fine repeating it here, but it has me wondering about re-use between front- and backend code in the future.

beechnut replied:
That...is a great question and definitely worth exploring. I opened #373 to dig into this further.

neilmb reviewed (dismissed):
LGTM. Lots of questions, but nothing that needs to stand in the way.

Commit:
Connect frontend and backend

- Connects Angular app and Express API locally and in 18F AWS Sandbox
- Pins Dockerfiles to NodeJS v14
- Adds Prettier to format frontend code in CI
- Rearranges CI config to run (fast) code analysis first, before slower work
- Converts demo app to use grants

neilmb reviewed (approved):
LGTM. Glad to see Prettier at work in the source code and worked well locally with `docker-compose up`.

jayrbergjr commented:
@beechnut I approve, I can't figure out how to check my request.  :(
